File: frontend/lib/models/meeting.dart
import 'dart:convert';

class Meeting {
  final String id;
  final DateTime begin;
  final DateTime end;
  final String place;
  final String title;
  final String kind;
  final List<String> communicationsId;
  final String? minute;
  final MeetingParticipants participants;

  Meeting(
      {required this.id,
      required this.begin,
      required this.end,
      required this.place,
      required this.title,
      required this.kind,
      required this.communicationsId,
      this.minute,
      required this.participants});

  factory Meeting.fromJson(Map<String, dynamic> json) {
    return Meeting(
        id: json['id'],
        begin: DateTime.parse(json['begin']),
        end: DateTime.parse(json['end']),
        place: json['place'],
        minute: json['minute'],
        title: json['title'],
        kind: json['kind'],
        communicationsId: List.from(json['communications']),
        participants: MeetingParticipants.fromJson(json['participants']));
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'begin': begin,
        'end': end,
        'place': place,
        'minute': minute,
        'title': title,
        'kind': kind,
        'communications': communicationsId,
        'participants': participants.toJson()
      };

  @override
  String toString() {
    return json.encode(this.toJson());
  }
}

class MeetingParticipants {
  final List<String>? membersIds;
  final List<String>? companyRepIds;

  MeetingParticipants({this.membersIds, this.companyRepIds});

  factory MeetingParticipants.fromJson(Map<String, dynamic> json) {
    return MeetingParticipants(
        membersIds: json['members'], companyRepIds: json['companyReps']);
  }

  Map<String, dynamic> toJson() =>
      {'members': membersIds, 'companyReps': companyRepIds};

  @override
  String toString() {
    return json.encode(this.toJson());
  }
}

File: backend/src/models/meeting.go
package models

import (
	"errors"
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

type MeetingKind string
type MeetingParticipantKind string

const (
	EventMeeting   MeetingKind = "EVENT"
	TeamMeeting    MeetingKind = "TEAM"
	CompanyMeeting MeetingKind = "COMPANY"
)

const (
	MemberParticipant     MeetingParticipantKind = "MEMBER"
	CompanyRepParticipant MeetingParticipantKind = "COMPANYREP"
)

func (mk *MeetingKind) Parse(kind string) error {

	var newMeetingKind MeetingKind

	switch kind {

	case string(EventMeeting):
		newMeetingKind = EventMeeting
		break

	case string(TeamMeeting):
		newMeetingKind = TeamMeeting
		break

	case string(CompanyMeeting):
		newMeetingKind = CompanyMeeting
		break

	default:
		return errors.New("invalid kind")

	}

	*mk = newMeetingKind
	return nil
}

func (mk *MeetingParticipantKind) Parse(participantKind string) error {

	var newMeetingParticipantKind MeetingParticipantKind

	switch participantKind {

	case string(MemberParticipant):
		newMeetingParticipantKind = MemberParticipant
		break

	case string(CompanyRepParticipant):
		newMeetingParticipantKind = CompanyRepParticipant
		break

	default:
		return errors.New("invalid type of participant")

	}

	*mk = newMeetingParticipantKind
	return nil
}

type MeetingParticipants struct {

	// Members is an array of _id of Member (see models.Member).
	Members []primitive.ObjectID `json:"members" bson:"members"`

	// CompanyReps is an array of _id of CompanyRep (see models.CompanyRep).
	CompanyReps []primitive.ObjectID `json:"companyReps" bson:"companyReps"`
}

// Meeting stores informations abouts meetings with the SINFO team,
// or meetings scheduled with companies.
type Meeting struct {

	// Meeting's ID (_id of mongodb).
	ID primitive.ObjectID `json:"id" bson:"_id"`

	// The title of the meeting
	Title string `json:"title" bson:"title"`

	// Type of the meeting
	Kind MeetingKind `json:"kind" bson:"kind"`

	Begin time.Time `json:"begin" bson:"begin"`
	End   time.Time `json:"end" bson:"end"`

	// Place where the meeting is being held.
	Place string `json:"place" bson:"place"`

	// A written record of the meeting. This is typically a URL to where the
	// actual document is being stored.
	// "Ata" in portuguese.
	Minute string `json:"minute" bson:"minute"`

	// Communications is an array of _id of Communication (see models.Communication).
	Communications []primitive.ObjectID `json:"communications" bson:"communications"`

	Participants MeetingParticipants `json:"participants" bson:"participants"`
}
